### PR TITLE
fix(framework): bug with duplicate step names

### PR DESCRIPTION
**Pull request by denis-kralj-novu**
- Move error detection to action endpoints
- Remove throwing on ID collision, log error instead 
- Add tests

### CODE REVIEW

File: packages/framework/src/handler.ts [hunk 1]
```diff
@@ -176,8 +176,21 @@ export class NovuRequestHandler<Input extends any[] = any[], Output = any> {
...
+        const errors = this.client.getErrors();
+
+        if (errors.length !== 0) {
+          return this.createResponse(HttpStatusEnum.UNPROCESSABLE_ENTITY, errors);
+        }
+
+        return this.triggerAction({ workflowId, ...body })();
+      },
       [PostActionEnum.EXECUTE]: async () => {
+        const errors = this.client.getErrors();
```
Comment (denis-kralj-novu) [inline packages/framework/src/handler.ts:189]: ❓ **question:** ‏ what is the usecase for execute? I'm not sure how it gets used in the studio and the bridge api client doesn't seem to use the action.

File: packages/framework/src/handler.ts [hunk 2]
```diff
@@ -223,17 +241,27 @@ export class NovuRequestHandler<Input extends any[] = any[], Output = any> {
...
       },
       [GetActionEnum.HEALTH_CHECK]: async () => {
-        const result = await this.client.healthCheck();
+        const result = this.client.healthCheck();
 
         return this.createResponse(HttpStatusEnum.OK, result);
       },
       [GetActionEnum.CODE]: async () => {
-        const result = await this.client.getCode(workflowId, stepId);
+        const errors = this.client.getErrors();
```
Comment (denis-kralj-novu) [inline packages/framework/src/handler.ts:259]: ❓ **question:** ‏ same as execute, what is the usecase for code? I'm not sure how it gets used in the studio and the bridge api client doesn't seem to use the action.

File: packages/framework/src/client.ts [hunk 3]
```diff
@@ -149,6 +150,47 @@ export class Client {
     };
   }
 
+  public getErrors(): string[] {
+    const errors: string[] = [];
+
+    const workflowIds = this.discoveredWorkflows.map((discoveredWorkflow) => discoveredWorkflow.workflowId);
+
+    const workflowOccuranceMap = this.getOccuranceMap(workflowIds);
```
Comment (SokratisVidros) [inline packages/framework/src/client.ts:158]: **suggestion:** ‏ We can do this check in a single pass with an assert function that will throw an error in the first duplicate detection using a reduce function. The accumulator is the occuranceMap.

File: packages/framework/src/handler.ts [hunk 4]
```diff
@@ -176,8 +176,21 @@ export class NovuRequestHandler<Input extends any[] = any[], Output = any> {
...
+        if (errors.length !== 0) {
+          return this.createResponse(HttpStatusEnum.UNPROCESSABLE_ENTITY, { errors });
+        }
+
+        return this.triggerAction({ workflowId, ...body })();
+      },
       [PostActionEnum.EXECUTE]: async () => {
+        const errors = this.client.getErrors();
+
+        if (errors.length !== 0) {
```
Comment (SokratisVidros) [inline packages/framework/src/handler.ts:191]: Let's generalize error handling across all action types to make the code dryer. Every byte counts in the framework SDK.

File: packages/framework/src/handler.ts (hunk 2 above)
Comment (SokratisVidros) [reply packages/framework/src/handler.ts:259]: cc: @rifont ☝️

File: packages/framework/src/handler.ts (hunk 1 above)
Comment (SokratisVidros) [reply packages/framework/src/handler.ts:189]: cc: @rifont ☝️

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

**Pull request closed without merging**